litestar/handlers/base.py

```diff
@@ -287,14 +339,31 @@ def resolve_dependencies(self) -> dict[str, Provide]:
             self._resolved_dependencies = {}
 
             for layer in self.ownership_layers:
-                for key, value in (layer.dependencies or {}).items():
-                    if not isinstance(value, Provide):
-                        value = Provide(value)
+                for key, provider in (layer.dependencies or {}).items():
+                    if not isinstance(provider, Provide):
+                        provider = Provide(provider)
+
                     self._validate_dependency_is_unique(
-                        dependencies=self._resolved_dependencies, key=key, provider=value
+                        dependencies=self._resolved_dependencies, key=key, provider=provider
                     )
-                    self._resolved_dependencies[key] = value
 
+                    if not getattr(provider, "signature_model", None):
+                        provider.signature_model = SignatureModel.create(
+                            dependency_name_set=self.dependency_name_set,
+                            fn=provider.dependency.value,
+                            parsed_signature=ParsedSignature.from_fn(
+                                unwrap_partial(provider.dependency.value), self.resolve_signature_namespace()
+                            ),
+                            data_dto=self.resolve_data_dto(),
+                            type_decoders=self.resolve_type_decoders(),
+                        )
+                    if provider.sync_to_thread and not is_async_callable(provider.dependency.value):
+                        provider.dependency.value = async_partial(provider.dependency.value)
+                        provider.has_sync_callable = False
+                    else:
+                        provider.has_sync_callable = not is_async_callable(provider.dependency.value)
+
+                    self._resolved_dependencies[key] = provider
```

Comment:
Why is this happening in here? Isn't this already being handled within `Provide`?

Reply:
Yes, it's better to move the logic into probably provide. And no, it's not happening there but happened in the handler.

Reply:
done